forc-pkg/src/pkg.rs

```diff
@@ -1758,7 +1854,24 @@ fn dep_to_source(pkg_path: &Path, dep: &Dependency) -> Result<Source> {
                 let canonical_path = path.canonicalize().map_err(|e| {
                     anyhow!("Failed to canonicalize dependency path {:?}: {}", path, e)
                 })?;
-                Source::Path(canonical_path)
+                // Check if path is a member of a workspace.
+                let workspace_manifest = canonical_path
+                    .parent()
+                    .and_then(|parent_dir| WorkspaceManifestFile::from_dir(parent_dir).ok());
+
+                if let Some(workspace_manifest) = workspace_manifest {
+                    // There is a workspace manifest in the parent directory.
+                    //
+                    // Check if that manifest declares this path as a member_path
+                    if workspace_manifest.is_member_path(&canonical_path)? {
+                        // This is a workspace member so the source should be inserted as `Member`
+                        Source::Member(canonical_path)
+                    } else {
+                        Source::Path(canonical_path)
+                    }
+                } else {
+                    Source::Path(canonical_path)
+                }
```

Comment:
This could also be written as:
```rust
match workspace_manifest {
    Some(ws) if ws.is_member_path(&canonical_path) => Source::Member(canonical_path),
    _ => Source::Path(canonical_path),
}
```